### ISSUE TITLE
IMAP caches are not used?

### ISSUE THREAD

**Issue by phmarek** [neomutt/neomutt#154]
Hi,

since ~2 months my mutt (debian 1.7.0) doesn't use the local caches correctly.

When re-visiting some IMAP folders it says "mailbox unchanged", but then starts to fetch all email headers afresh. Please see http://paste.lisp.org/display/326529 (manually translated a few messages from German; mailbox substituted by "*"); I didn't choose the "Drafts" folder!

Thanks for the help.

**Comment by P-EB** [comment]
"Mailbox unchanged" is about the mailbox you're leaving, not the one you're fetching.

What happens if you try to go to another folder and back to the current one immediately? Does it fetch everything without using cache?

**Comment by phmarek** [comment]
oh, okay.

Well, sometimes it fetches _everything_, yes. Can't say when or why yet.

**Comment by P-EB** [comment]
I guess we'll have to dig into the code. :)

**Comment by P-EB** [comment]
This seems related to issue #160

**Comment by gahr** [comment]
When selecting an IMAP folder, mutt fetches the UID and FLAGS attributes. Isn't this what you're seeing?

**Comment by gahr** [comment]
@phmarek sorry your paste has expired. Could you please post your .muttdebug0 when running mutt with -d3?

**Comment by toogley** [comment]
@phmarek ... and to prevent that the paste is lost again in the future, could you add it to your post? thanks.

**Comment by phmarek** [comment]
Hmmm, I can't see any duplicated `imap_read_literal` right now. Or perhaps I misread the log file?
Most of the time I don't have debugging active; is there a way to switch the debug log on while having mutt running?

Having the logfile auto-expire was on purpose, in case I left too much information in it ;)

**Comment by flatcap** [comment]
> is there a way to switch the debug log on while having mutt running?

No, but that's a reasonable thing to be able to do.  (added to the list)

**Comment by flatcap** [comment]
This issue has been quiet for a while, now.
Hopefully this means there haven't been any further problems.
Intermittent problems, like this, are hard to debug, but each release of NeoMutt adds more fixes and checks.
If the problem reoccurs, please re-open this issue. 

As for your question about logging.
It's now possible to dynamically enable logging with `$debug_level` and `$debug_file`

**Comment by oblitum** [comment]
I'm also suffering from the same issue, it started to happen after I've changed my `tmpdir` to `/tmp/mutt/`. If I change it back to `/tmp` by removing the setting it starts to use the cache normally. With `/tmp/mutt` it is fetching all headers most of the time I start mutt.

**Comment by mschilli87** [comment]
@oblitum: Did you ensure that `/tmp/mutt` *exists*? AFAIR, (neo)mutt doesn't check/create it. Thus, it tries to write the cache files but (silently) fails to do so. So probably `mkdir /tmp/mutt` will be enough to make it work.

**Comment by oblitum** [comment]
@mschilli87 just tried, same behavior.

**Comment by oblitum** [comment]
OK I've fixed my problem with:

```diff
set mail_check       = 60
 set timeout          = 10
 set imap_keepalive   = 150
 set postponed        = '+[GMail]/Drafts'
-set header_cache     = ~/.mutt/cache/headers
-set message_cachedir = ~/.mutt/cache/bodies
+set header_cache     = "~/.mutt/cache/$MUTT_ACCOUNT-headers"
+set message_cachedir = "~/.mutt/cache/$MUTT_ACCOUNT-bodies"
 set certificate_file = "~/.mutt/certificates"
 set move             = no
 set sort             = 'threads'
 set sort_aux         = 'reverse-last-date-received'
```
I make use of multiple accounts by multiple neomutt instances, each in its own tmux window. I think using the sample cache file/dir for all was the problem. NeoMutt debug level didn't help at all.

**Comment by oblitum** [comment]
I don't know why the problem was less frequent before, it started to happen always after I've deleted my cache directory and I thought it was due to  the `tmpdir` setting or other thing.

**Comment by oblitum** [comment]
Actually, this is the actual fix:

```diff
-set header_cache     = ~/.mutt/cache/headers
-set message_cachedir = ~/.mutt/cache/bodies
+set header_cache     = ~/.mutt/cache/headers/
+set message_cachedir = ~/.mutt/cache/bodies/
```
The problems started when I deleted my cache dir. I was formerly using mutt for sending emails through smtp but changed to msmtp for that, so mutt was only accessing imap. On my first setup, I think due to using smtp maybe, I don't know, but mutt had created `~/.mutt/cache/headers` as a directory, but after I've deleted the cache dir e restarted the multiple instances with new imap only setup, it created `~/.mutt/cache/headers` as a single file, and the multiple instances were using it for all accounts. Adding `/` to the end of the path makes mutt create it as a directory, despite `~/.mutt/cache` not existing at first. Then it starts to create automatically named imap folders correctly at `~/.mutt/cache/headers/`.
